### PR TITLE
Fix android webview onMessage should return baseEvent

### LINKED ISSUE #119: WebView `onMessage` behaviour is not same in both OS
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
System:
OS: macOS 10.13.1
CPU: x64 Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz
Memory: 500.23 MB / 16.00 GB
Shell: 3.2.57 - /bin/zsh
Binaries:
Node: 8.11.4 - /usr/local/bin/node
Yarn: 1.10.1 - /usr/local/bin/yarn
npm: 6.2.0 - /usr/local/bin/npm
Watchman: 4.9.0 - /usr/local/bin/watchman
SDKs:
iOS SDK:
Platforms: iOS 12.0, macOS 10.14, tvOS 12.0, watchOS 5.0
Android SDK:
Build Tools: 23.0.1, 25.0.2, 27.0.3, 28.0.0, 28.0.1
API Levels: 20, 23, 24, 25, 26, 27, 28
IDEs:
Android Studio: 3.1 AI-173.4819257
Xcode: 10.0/10A255 - /usr/bin/xcodebuild
npmPackages:
react: ^16.6.0-alpha.8af6728 => 16.6.0-alpha.8af6728
react-native: ^0.57.4 => 0.57.4
npmGlobalPackages:
create-react-native-app: 2.0.2
react-native-cli: 2.0.1
react-native-git-upgrade: 0.2.7

## Description
Android & iOS webview `onMessage` behaviour is not same, iOS will return `baseEvent` but Android wont.

## Reproducible Demo
https://snack.expo.io/@jl9404/webview-demo

### PR DESCRIPTION
**Pull request by jl9404**
Fixes #119 

Following https://snack.expo.io/@jl9404/webview-demo, after this fix, android webview should be able to receive bridge event from webview with baseEvent (canGoBack, canGoFoward, etc...), this fix will sync the behaviour between Android and iOS

### CODE REVIEW

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java [hunk 1]
```diff
@@ -369,7 +369,11 @@ public void onReceiveValue(String value) {
     }
 
     public void onMessage(String message) {
-      dispatchEvent(this, new TopMessageEvent(this.getId(), message));
+      if (mRNCWebViewClient != null) {
+        WritableMap eventData = mRNCWebViewClient.createWebViewEvent(this, this.getUrl());
+        eventData.putString("data", message);
+        dispatchEvent(this, new TopMessageEvent(this.getId(), eventData));
```
Comment (jamonholmgren) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Do we want this line outside of the conditional? If `mRNCWebViewClient` is null, it won't get dispatched.

Comment (jl9404) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: anyway to guarantee `mRNCWebViewClient` is not `null`? from the previous code, it seems that `mRNCWebViewClient` wont be `null`

Comment (jamonholmgren) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: @jl9404 If `mRNCWebViewClient` is null, then do you not want to dispatch that event, then? I thought lines 373-374 were the only ones that needed to check if it was null, but I could be wrong here.

Comment (Titozzz) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: mRNCWebViewClient has not reason to be null at this point in the code, and I feel like I'd rather miss an event than crash if it happened to be null ^^'

Comment (jamonholmgren) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Not to belabor the point, but I was wondering if we wanted to do this, which would both protect against null and not miss an event:

```js
    public void onMessage(String message) {
      if (mRNCWebViewClient != null) {
        WritableMap eventData = mRNCWebViewClient.createWebViewEvent(this, this.getUrl());
        eventData.putString("data", message);
      }
      dispatchEvent(this, new TopMessageEvent(this.getId(), eventData));
    }
```

Comment (Titozzz) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: @jamonholmgren you can't access eventData in the code your wrote

Comment (jamonholmgren) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Oops, you're totally right. 🤦‍♂️

**Comment by jamonholmgren** [comment]
@jl9404 Can you resolve conflicts? Once that's done we can merge and ship this.

**Comment by jamonholmgren** [comment]
@jl9404 Will you have time to merge the conflicts on this?

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java (hunk 1 above)
Comment (keenubee) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: @Titozzz @jamonholmgren Maybe
```js
public void onMessage(String message) {
  WritableMap eventData;
  if (mRNCWebViewClient != null) {
    eventData = mRNCWebViewClient.createWebViewEvent(this, this.getUrl());
  } else {
    eventData = Arguments.createMap();
  }
  eventData.putString("data", message);
  dispatchEvent(this, new TopMessageEvent(this.getId(), eventData));
}
```
?

Comment (keenubee) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: BTW `this.getURL()` throws
`java.lang.Throwable: A WebView method was called on thread 'JavaBridge'. All WebView methods must be called on the same thread. (Expected Looper Looper (main, tid 2) {eead3fd} called on Looper (JavaBridge, tid 1541) {63b1022}, FYI main Looper is Looper (main, tid 2) {eead3fd})`

Comment (keenubee) [reply android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: Working code:
```js
public void onMessage(String message) {
  if (mRNCWebViewClient != null) {
    WebView wv = this;
    wv.post(new Runnable() {
      @Override
      public void run() {
        WritableMap data = mRNCWebViewClient.createWebViewEvent(wv, wv.getUrl());
        data.putString("data", message);
        dispatchEvent(wv, new TopMessageEvent(wv.getId(), data));
      }
    });
  } else {
    WritableMap eventData = Arguments.createMap();
    eventData.putString("data", message);
    dispatchEvent(this, new TopMessageEvent(this.getId(), eventData));
  }
}
```
But all data inside `data` are empty in my case.

Comment (keenubee) [inline android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java]: @Titozzz Can you please check these changes?
```js
public void onMessage(String message) {
  if (mRNCWebViewClient != null) {
    WebView wv = this;
    wv.post(new Runnable() {
      @Override
      public void run() {
        WritableMap data = mRNCWebViewClient.createWebViewEvent(wv, wv.getUrl());
        data.putString("data", message);
        dispatchEvent(wv, new TopMessageEvent(wv.getId(), data));
      }
    });
  } else {
    WritableMap eventData = Arguments.createMap();
    eventData.putString("data", message);
    dispatchEvent(this, new TopMessageEvent(this.getId(), eventData));
  }
}
```
Currently your code throws
```
java.lang.Throwable: A WebView method was called on thread 'JavaBridge'. All WebView methods must be called on the same thread. (Expected Looper Looper (main, tid 2) {eead3fd} called on Looper (JavaBridge, tid 1541) {63b1022}, FYI main Looper is Looper (main, tid 2) {eead3fd})
```
exception inside `this.getUrl()`.

Also i've noticed that my changes return empty `data` in onMessage handler.

To resolve the conflict you can use this code for `TopMessageEvent.kt`:
```
package com.reactnativecommunity.webview.events

import com.facebook.react.bridge.WritableMap
import com.facebook.react.uimanager.events.Event
import com.facebook.react.uimanager.events.RCTEventEmitter

/**
 * Event emitted when there is an error in loading.
 */
class TopMessageEvent(viewId: Int, private val mEventData: WritableMap) :
    Event<TopMessageEvent>(viewId) {
    companion object {
        const val EVENT_NAME = "topMessage"
    }

    override fun getEventName(): String = EVENT_NAME

    override fun canCoalesce(): Boolean = false

    override fun getCoalescingKey(): Short = 0

    override fun dispatch(rctEventEmitter: RCTEventEmitter) =
        rctEventEmitter.receiveEvent(viewTag, eventName, mEventData)
}
```

**Comment by BukhariMuslim** [comment]
It's 7 months passed. Any update on this?

**Comment by jamonholmgren** [comment]
@BukhariMuslim I lack the context to know the answer to the questions that @keenubee asked above. Do you know enough to weigh in?

**Comment by BukhariMuslim** [comment]
I don't have enough context to weigh in. So I'm just doin some quick search with the error message and got this. [StactOverflow link: WebView Methods on Same Thread Error](https://stackoverflow.com/questions/22607657/webview-methods-on-same-thread-error)
I'm not really understand but it might help.

**Comment by Titozzz** [comment]
If no updates are made to this code, I'll close the PR.

**Pull request closed without merging**